File: Backend/application/service/product_service.py
from application.dto.productDTO import ProductDTO
from application.model.models import Product
from application.repository.product_repository import ProductRepository


class ProductService:
    @staticmethod
    def create_product(**data):
        product_DTO = ProductDTO()
        product_data = product_DTO.load(data)

        product_new = Product(name=product_data.get('name'),
                              quantity_min=product_data.get('quantity_min'),
                              quantity_max=product_data.get('quantity_max'),
                              stock=product_data.get('stock'),
                              image_url=product_data.get('image_url')
                              )

        return ProductRepository.add_product(product_new)

    @staticmethod
    def delete_product(product_id):
        return ProductRepository.delete_product(product_id)

    @staticmethod
    def update_product(product_id, data):
        return ProductRepository.update_product(product_id, data)

    @staticmethod
    def get_product_by_id(product_id):
        return ProductRepository.find_product_by_id(product_id)

    @staticmethod
    def get_product_by_name(product_name):
        return ProductRepository.find_product_by_name(product_name)

    @staticmethod
    def get_all_products():
        products = ProductRepository.find_all_products()
        return [product.repr() for product in products]

File: Frontend/src/app/service/product/product.service.ts
import { Injectable } from '@angular/core';
import {HttpClient, HttpHeaders} from "@angular/common/http";
import {BehaviorSubject, Observable} from "rxjs";
import {Product} from "../../model/product";
import {API_CONFIG} from "../../app.config";

@Injectable({
  providedIn: 'root'
})
export class ProductService {
  baseURL: string = `${API_CONFIG.baseIP}/product`;
  ownerDataStream: any;

  constructor(private httpClient:HttpClient) {
    this.ownerDataStream = new BehaviorSubject<any>(null);
  }

  // header - pentru header-ul mesajului
  // metoda de get - URL + id (path variable), header
  getProductById(id: any): Observable<Product> {
    let header = new HttpHeaders()
      .set('Content-Type', 'application/json')
    return this.httpClient.get<Product>(this.baseURL + "/getProductById/" + id, {headers: header})
  }

  // header - pentru header-ul mesajului
  // metoda de get - URL + name (path variable), header
  getProductByName(name: any): Observable<Product> {
    let header = new HttpHeaders()
      .set('Content-Type', 'application/json')
    return this.httpClient.get<Product>(this.baseURL + "/getProductByName/" + name, {headers: header})
  }

  // header - pentru header-ul mesajului
  // metoda de get - URL, header
  // folosita in pagina de produse
  getAllProducts(): Observable<Product[]> {
    let header = new HttpHeaders()
      .set('Content-Type', 'application/json')
    return this.httpClient.get<Product[]>(this.baseURL + "/getAllProducts", {headers: header})
  }

  // header - pentru header-ul mesajului
  // infos - name, quantity_min, quantity_max, stock, image_url
  // metoda de post - URL, body, header
  addProduct(name: any, quantity_min: any, quantity_max: any, stock: any, image_url: any){
    let header = new HttpHeaders()
      .set('Content-Type', 'application/json')
    let infos = {name: name, quantity_min: quantity_min, quantity_max: quantity_max, stock: stock, image_url: image_url}
    return this.httpClient.post(this.baseURL + "/addProduct", JSON.stringify(infos), {headers: header})
  }

  // header - pentru header-ul mesajului
  // infos - name, quantity_min, quantity_max, stock, image_url
  // metoda de put - URL, body, header
  updateProduct(id: any, name: any, quantity_min: any, quantity_max: any, stock: any, image_url: any){
    let header = new HttpHeaders()
      .set('Content-Type', 'application/json')
    let infos = {name: name, quantity_min: quantity_min, quantity_max: quantity_max, stock: stock, image_url: image_url}
    return this.httpClient.put(this.baseURL + "/updateProduct/" + id, JSON.stringify(infos), {headers: header})
  }

  // header - pentru header-ul mesajului
  // metoda de delete - URL + id (path variable), header
  deleteProduct(id: any){
    let header = new HttpHeaders()
      .set('Content-Type', 'application/json')
    return this.httpClient.delete(this.baseURL + "/deleteProduct/" + id, {headers: header})
  }

}
